### PR TITLE
[reactive-element] Make context decorators work with standard decorators

### PR DESCRIPTION
**Pull request by justinfagnani**
Part of https://github.com/lit/lit/issues/4144

One thing to check on here is how optional fields are translated to auto-accessors and whether or not we have the types correct, and had them correct prior to this PR. The exist tests include a case of an optional field decorated with a non-optional context type. That doesn't exactly work with accessors, which can't be optional, and so have to include `undefined` in their type to get the same effect. This difference is that this setup requires the context include `undefined`, whereas optional fields _don't_. The actual type of optional fields is the same though, so it seems like something was wrong in the original types.

### CODE REVIEW

**Review by rictic — COMMENTED** [review]
This is a small subset of the tests in https://github.com/lit/lit/pull/4009/files

I can land that PR after this one and land all of the tests as part of that

**Comment by justinfagnani** [comment]
I didn't copy over all of the tests because I'm not entirely happy with how the tests are organized right now. Ideally I would have just copied over the decorator tests, but there aren't tests for just the decorators. There are tests for scenarios, and some components of the tests directly use controllers or use decorators.

It's a little tricky because of the two-part nature of the protocol. It's easier to test a consumer if you have a provider and vice-versa. I don't think it's that great to reimplement half of the context protocol with the possibility of mistakes there. It would be great if there were common protocol interop tests. But even with a combined comsumer/provider test, that could use the controllers only and then we could have isolated and smaller tests for decorators.

File: packages/labs/context/src/lib/decorators/consume.ts [hunk 1]
```diff
@@ -47,28 +46,60 @@ export function consume<ValueType>({
...
+      );
+    }
+  }) as ConsumeDecorator<ValueType>;
 }
 
-type ConsumerDecorator<ValueType> = {
-  <K extends PropertyKey, Proto extends ReactiveElement>(
+/**
+ * Generates a public interface type that removes private and protected fields.
+ * This allows accepting otherwise compatible versions of the type (e.g. from
```
[the remark this answers is not in the document]
Comment (justinfagnani) [reply packages/labs/context/src/lib/decorators/consume.ts:84]: done

File: packages/labs/context/src/lib/decorators/provide.ts [hunk 2]
```diff
@@ -48,15 +47,41 @@ export function provide<ValueType>({
...
-      ctor.addInitializer((element: ReactiveElement): void => {
-        controllerMap.set(element, new ContextProvider(element, {context}));
+  return (<C extends ReactiveElement, V extends ValueType>(
+    protoOrTarget: ClassAccessorDecoratorTarget<C, V>,
+    nameOrContext: PropertyKey | ClassAccessorDecoratorContext<C, V>
+  ) => {
+    // Map of instances to controllers
+    const controllerMap = new WeakMap();
+    if (typeof nameOrContext === 'object') {
+      nameOrContext.addInitializer(function (this: C) {
```
Comment (rictic) [inline packages/labs/context/src/lib/decorators/provide.ts:57]: Here and for consume: let's add a comment saying essentially "this is the standard decorators branch" and "this is the legacy experimental decorators branch"

File: packages/reactive-element/src/test/decorators/property_test.ts [hunk 3]
```diff
@@ -160,10 +160,10 @@ suite('@property', () => {
 
     // Check initial values
     assert.equal(el._foo, undefined);
-    assert.isFalse(el.updatedProperties!.has('foo'));
+    assert.equal(el.updatedProperties!.get('foo'), undefined);
```
Comment (rictic) [inline packages/reactive-element/src/test/decorators/property_test.ts:163]: Looks like a merge error.

**Review by rictic — APPROVED** [review]

File: packages/labs/context/src/lib/decorators/provide.ts (hunk 2 above)
Comment (justinfagnani) [reply packages/labs/context/src/lib/decorators/provide.ts:57]: done

**Pull request merged**